File: frontend/src/v2/api/channels.ts
// frontend/src/v2/api/channels.ts
//
// 渠道域 API 层。禁止页面层直接调 axios，统一走 apiClient。
// 后端契约：app/interfaces/api/v1/channels.py
//           app/domain/entities/config/channel.py

import { apiClient } from './client'
import type { PaginatedResponse } from './types'

// ============================================================================
// 类型定义（snake_case 与后端 wire 格式一致）
// ============================================================================

/** 后端实际支持的渠道类型（align: app/domain/entities/config/channel.py ChannelType） */
export type ChannelType = 'feishu' | 'email' | 'webhook' | 'oss'

/**
 * 渠道配置结构按类型不同：
 * feishu:  { chat_id?, webhook_url?, message_template? }
 * email:   { recipients, subject_template?, body_template? }
 * webhook: { url, method?, headers?, body_template? }
 * oss:     { bucket, path_template?, filename_template? }
 */
export type ChannelConfig = Record<string, unknown>

export interface Channel {
  id: number
  name: string
  channel_type: ChannelType
  description: string | null
  config: ChannelConfig
  enabled: boolean
  created_by: string | null
  created_at: string | null
  updated_at: string | null
}

export interface ChannelListParams {
  channel_type?: ChannelType
  enabled?: boolean
  page?: number
  page_size?: number
}

export interface CreateChannelPayload {
  name: string
  channel_type: ChannelType
  config: ChannelConfig
  description?: string
  enabled?: boolean
}

export interface UpdateChannelPayload {
  name?: string
  config?: ChannelConfig
  description?: string
  enabled?: boolean
}

// ============================================================================
// API 函数
// ============================================================================

export async function listChannels(
  params: ChannelListParams = {},
): Promise<PaginatedResponse<Channel>> {
  const res = await apiClient.get<{ data: PaginatedResponse<Channel> }>('/channels', { params })
  return res.data.data
}

export async function getChannel(id: number): Promise<Channel> {
  const res = await apiClient.get<{ data: Channel }>(`/channels/${id}`)
  return res.data.data
}

export async function createChannel(payload: CreateChannelPayload): Promise<Channel> {
  const res = await apiClient.post<{ data: Channel }>('/channels', payload)
  return res.data.data
}

export async function updateChannel(id: number, payload: UpdateChannelPayload): Promise<Channel> {
  const res = await apiClient.put<{ data: Channel }>(`/channels/${id}`, payload)
  return res.data.data
}

export async function deleteChannel(id: number): Promise<void> {
  await apiClient.delete(`/channels/${id}`)
}

export async function enableChannel(id: number): Promise<Channel> {
  const res = await apiClient.post<{ data: Channel }>(`/channels/${id}/enable`)
  return res.data.data
}

export async function disableChannel(id: number): Promise<Channel> {
  const res = await apiClient.post<{ data: Channel }>(`/channels/${id}/disable`)
  return res.data.data
}

// ── 测试发送（P12）────────────────────────────────────────────────────────────
// 后端契约：POST /api/v1/channels/:id/test
//   body  : { message?: string }
//   data  : ChannelTestResult（字段与 channel_service.test_channel 返回值对齐）

export interface ChannelTestResult {
  ok: boolean
  channel_type: ChannelType
  latency_ms: number
  /** HTTP 状态码（feishu / webhook 类型可用，email / oss 为 null） */
  status_code: number | null
  /** 人类可读的诊断信息 */
  detail: string
  /** 失败原因（成功时为 null） */
  error: string | null
  /** true 表示仅做配置校验，未实际发送消息 */
  dry_run: boolean
}

export interface TestChannelPayload {
  /** 可选：自定义测试消息文本（飞书 / Webhook 使用） */
  message?: string
}

export async function testChannel(
  id: number,
  payload: TestChannelPayload = {},
): Promise<ChannelTestResult> {
  const res = await apiClient.post<{ data: ChannelTestResult }>(
    `/channels/${id}/test`,
    payload,
  )
  return res.data.data
}

File: app/interfaces/api/v1/channels.py
"""
渠道管理 API
"""
from flask import Blueprint, request, g

from app.di.container import get_container
from app.interfaces.api.middleware.auth import require_auth
from app.shared.response import success, created
from app.shared.utils.logger import get_logger

logger = get_logger(__name__)
bp = Blueprint('channels_api_v1', __name__, url_prefix='/api/v1/channels')


def _get_service():
    return get_container().channel_service()


@bp.route('', methods=['GET'])
@require_auth
def list_channels():
    """查询渠道列表"""
    channel_type = request.args.get('channel_type')
    enabled = request.args.get('enabled')
    page = request.args.get('page', 1, type=int)
    page_size = request.args.get('page_size', 20, type=int)

    if enabled is not None:
        enabled = enabled.lower() == 'true'

    service = _get_service()
    result = service.list_channels(
        channel_type=channel_type,
        enabled=enabled,
        page=page,
        page_size=page_size
    )
    return success(data=result)


@bp.route('', methods=['POST'])
@require_auth
def create_channel():
    """创建渠道"""
    data = request.get_json()
    user_id = g.user_id

    service = _get_service()
    channel = service.create_channel(
        name=data['name'],
        channel_type=data['channel_type'],
        config=data.get('config', {}),
        description=data.get('description'),
        created_by=user_id,
        enabled=data.get('enabled', True)
    )
    return created(data=channel)


@bp.route('/<int:channel_id>', methods=['GET'])
@require_auth
def get_channel(channel_id: int):
    """获取渠道详情"""
    service = _get_service()
    channel = service.get_channel(channel_id)
    return success(data=channel)


@bp.route('/<int:channel_id>', methods=['PUT'])
@require_auth
def update_channel(channel_id: int):
    """更新渠道"""
    data = request.get_json()

    service = _get_service()
    channel = service.update_channel(
        channel_id=channel_id,
        name=data.get('name'),
        config=data.get('config'),
        description=data.get('description'),
        enabled=data.get('enabled')
    )
    return success(data=channel)


@bp.route('/<int:channel_id>', methods=['DELETE'])
@require_auth
def delete_channel(channel_id: int):
    """删除渠道"""
    service = _get_service()
    service.delete_channel(channel_id)
    return success()


@bp.route('/<int:channel_id>/enable', methods=['POST'])
@require_auth
def enable_channel(channel_id: int):
    """启用渠道"""
    service = _get_service()
    channel = service.enable_channel(channel_id)
    return success(data=channel)


@bp.route('/<int:channel_id>/disable', methods=['POST'])
@require_auth
def disable_channel(channel_id: int):
    """禁用渠道"""
    service = _get_service()
    channel = service.disable_channel(channel_id)
    return success(data=channel)


@bp.route('/<int:channel_id>/test', methods=['POST'])
@require_auth
def test_channel(channel_id: int):
    """测试渠道连通性

    可选 body: ``{ "message": "自定义测试消息" }``

    响应: ``{ ok, channel_type, latency_ms, status_code, detail, error, dry_run }``
    """
    data = request.get_json(silent=True) or {}
    message = data.get('message')

    service = _get_service()
    result = service.test_channel(channel_id, message=message)
    return success(data=result)
